File: libs/core/src/symbols/namespace_symbol.cpp
/**
 * @file namespace_symbol.cpp
 * @author Francesco Corso
 * @date 8/8/2026
 * @brief
 */
module prism.core:symbols.namespace_symbol.impl;

import :symbols.namespace_symbol;

namespace prism
{
    Optional<const NamespaceSymbol &> NamespaceSymbol::get_nested_namespace(const Name name) const
    {
        for (const auto symbol : members(name))
        {
            if (auto ns = symbol->as<NamespaceSymbol>(); ns.has_value())
                return ns;
        }

        return std::nullopt;
    }

    void NamespaceSymbol::write_display_string(TextWriter &writer) const
    {
        if (const auto owner = containing_namespace(); owner.has_value() && !owner->is_global())
        {
            owner->write_display_string(writer);
            writer.write("::");
        }

        writer.write(name().as_string_view());
    }
} // namespace prism

File: src/Prism.Core/Symbols/NamespaceSymbol.cs
using Prism.Core.Compiling;
using Prism.Core.Utils;
using ZLinq;

namespace Prism.Core.Symbols;

public enum NamespaceKind
{
    Assembly,
    Compilation,
}

public abstract class NamespaceSymbol : MemberContainerSymbol
{
    private protected NamespaceSymbol(string name, Symbol? containingSymbol)
        : base(name, containingSymbol) { }

    public sealed override TypeSymbol? ContainingType => null;

    public bool IsGlobal => Name == CommonNames.Global;

    public sealed override bool IsImplicitlyDeclared => IsGlobal;

    public abstract NamespaceKind NamespaceKind { get; }

    public abstract Compilation? ContainingCompilation { get; }

    public NamespaceSymbol? GetNestedNamespace(string name)
    {
        return GetMembers(name).AsValueEnumerable().OfType<NamespaceSymbol>().FirstOrDefault();
    }

    public sealed override void WriteDisplayString(TextWriter writer)
    {
        if (ContainingNamespace is { IsGlobal: false } owner)
        {
            owner.WriteDisplayString(writer);
            writer.Write("::");
        }

        writer.Write(Name);
    }
}
